File: frontend/src/lib/api/income.ts
import { api } from '@/lib/auth';

export interface IncomeSource {
  id: string;
  name: string;
  description?: string;
  person_id?: string;
  income_type: 'recurring' | 'one_time' | 'project_based' | 'commission' | 'investment' | 'other';
  category: string;
  expected_amount: number;
  currency: string;
  is_recurring: boolean;
  recurrence_type?: 'daily' | 'weekly' | 'monthly' | 'quarterly' | 'yearly';
  recurrence_interval?: number;
  start_date: string;
  end_date?: string;
  is_active: boolean;
  notes?: string;
  total_received: number;
  total_expected: number;
  income_count: number;
  last_income_date?: string;
  next_expected_date?: string;
  created_by: string;
  created_at: string;
  updated_at?: string;
}

export interface IncomeSourceCreate {
  name: string;
  description?: string;
  person_id?: string;
  income_type: 'recurring' | 'one_time' | 'project_based' | 'commission' | 'investment' | 'other';
  category: string;
  expected_amount: number;
  currency?: string;
  is_recurring?: boolean;
  recurrence_type?: 'daily' | 'weekly' | 'monthly' | 'quarterly' | 'yearly';
  recurrence_interval?: number;
  start_date: string;
  end_date?: string;
  is_active?: boolean;
  notes?: string;
}

export interface IncomeSourceUpdate {
  name?: string;
  description?: string;
  person_id?: string;
  income_type?: 'recurring' | 'one_time' | 'project_based' | 'commission' | 'investment' | 'other';
  category?: string;
  expected_amount?: number;
  currency?: string;
  is_recurring?: boolean;
  recurrence_type?: 'daily' | 'weekly' | 'monthly' | 'quarterly' | 'yearly';
  recurrence_interval?: number;
  start_date?: string;
  end_date?: string;
  is_active?: boolean;
  notes?: string;
}

export interface IncomeSourceSummary {
  id: string;
  name: string;
  income_type: string;
  category: string;
  expected_amount: number;
  currency: string;
  total_received: number;
  total_expected: number;
  income_count: number;
  last_income_date?: string;
  next_expected_date?: string;
  status: string;
  person_name?: string;
  completion_rate: number;
}

export interface IncomeRecord {
  id: string;
  income_source_id: string;
  amount: number;
  currency: string;
  description: string;
  expected_date: string;
  actual_date?: string;
  status: 'planned' | 'invoiced' | 'paid' | 'overdue' | 'cancelled';
  invoice_number?: string;
  transaction_id?: string;
  bank_account_id?: string;
  payment_method?: string;
  late_fee: number;
  discount: number;
  tax_amount: number;
  net_amount: number;
  notes?: string;
  created_by: string;
  created_at: string;
  updated_at?: string;
}

export interface IncomeRecordCreate {
  income_source_id: string;
  amount: number;
  currency?: string;
  description: string;
  expected_date: string;
  actual_date?: string;
  status?: 'planned' | 'invoiced' | 'paid' | 'overdue' | 'cancelled';
  invoice_number?: string;
  transaction_id?: string;
  bank_account_id?: string;
  payment_method?: string;
  late_fee?: number;
  discount?: number;
  tax_amount?: number;
  notes?: string;
}

export interface IncomeStatistics {
  total_sources: number;
  active_sources: number;
  recurring_sources: number;
  total_expected_this_month: number;
  total_received_this_month: number;
  total_overdue: number;
  overdue_count: number;
  upcoming_week: number;
  upcoming_month: number;
  top_sources: Array<{
    id: string;
    name: string;
    total_received: number;
    income_type: string;
  }>;
  monthly_trends: Array<{
    month: string;
    total: number;
    count: number;
  }>;
  by_category: Array<{
    category: string;
    count: number;
    total_expected: number;
  }>;
  by_status: Array<{
    status: string;
    count: number;
    total_amount: number;
  }>;
}

export interface IncomeProjection {
  month: string;
  expected_total: number;
  confirmed_total: number;
  recurring_total: number;
  one_time_total: number;
  confidence_level: number;
}

export interface IncomeForecast {
  next_6_months: IncomeProjection[];
  yearly_projection: number;
  growth_rate: number;
  seasonality_factor: number;
  risk_level: string;
}

export const incomeApi = {
  // Income Sources
  async getIncomeSources(params?: {
    income_type?: string;
    category?: string;
    active_only?: boolean;
    limit?: number;
    skip?: number;
  }): Promise<IncomeSource[]> {
    const searchParams = new URLSearchParams();
    if (params?.income_type) searchParams.append('income_type', params.income_type);
    if (params?.category) searchParams.append('category', params.category);
    if (params?.active_only !== undefined) searchParams.append('active_only', params.active_only.toString());
    if (params?.limit) searchParams.append('limit', params.limit.toString());
    if (params?.skip) searchParams.append('skip', params.skip.toString());

    const queryString = searchParams.toString();
    const url = `/income/sources${queryString ? `?${queryString}` : ''}`;
    
    const response = await api.get(url);
    return response.data;
  },

  async getIncomeSourcesSummary(): Promise<IncomeSourceSummary[]> {
    const response = await api.get('/income/sources/summary');
    return response.data;
  },

  async getIncomeStatistics(): Promise<IncomeStatistics> {
    const response = await api.get('/income/statistics');
    return response.data;
  },

  async getIncomeSource(id: string): Promise<IncomeSource> {
    const response = await api.get(`/income/sources/${id}`);
    return response.data;
  },

  async createIncomeSource(data: IncomeSourceCreate): Promise<IncomeSource> {
    const response = await api.post('/income/sources', data);
    return response.data;
  },

  async updateIncomeSource(id: string, data: IncomeSourceUpdate): Promise<IncomeSource> {
    const response = await api.put(`/income/sources/${id}`, data);
    return response.data;
  },

  async deleteIncomeSource(id: string): Promise<void> {
    await api.delete(`/income/sources/${id}`);
  },

  // Income Records
  async getIncomeRecords(params?: {
    source_id?: string;
    status_filter?: string;
    start_date?: string;
    end_date?: string;
    limit?: number;
    skip?: number;
  }): Promise<IncomeRecord[]> {
    const searchParams = new URLSearchParams();
    if (params?.source_id) searchParams.append('source_id', params.source_id);
    if (params?.status_filter) searchParams.append('status_filter', params.status_filter);
    if (params?.start_date) searchParams.append('start_date', params.start_date);
    if (params?.end_date) searchParams.append('end_date', params.end_date);
    if (params?.limit) searchParams.append('limit', params.limit.toString());
    if (params?.skip) searchParams.append('skip', params.skip.toString());

    const queryString = searchParams.toString();
    const url = `/income/records${queryString ? `?${queryString}` : ''}`;
    
    const response = await api.get(url);
    return response.data;
  },

  async createIncomeRecord(data: IncomeRecordCreate): Promise<IncomeRecord> {
    const response = await api.post('/income/records', data);
    return response.data;
  },

  async getIncomeForecast(): Promise<IncomeForecast> {
    const response = await api.get('/income/forecast');
    return response.data;
  },

  // Utility functions
  getIncomeTypeLabel(type: string): string {
    const labels: Record<string, string> = {
      'recurring': 'Düzenli Gelir',
      'one_time': 'Tek Seferlik',
      'project_based': 'Proje Bazlı',
      'commission': 'Komisyon',
      'investment': 'Yatırım Getirisi',
      'other': 'Diğer'
    };
    return labels[type] || type;
  },

  getStatusLabel(status: string): string {
    const labels: Record<string, string> = {
      'planned': 'Planlandı',
      'invoiced': 'Faturalandı',
      'paid': 'Ödendi',
      'overdue': 'Gecikmiş',
      'cancelled': 'İptal'
    };
    return labels[status] || status;
  }
};

File: backend/app/models/income.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum

class IncomeType(str, Enum):
    RECURRING = "recurring"  # Düzenli gelir
    ONE_TIME = "one_time"   # Tek seferlik
    PROJECT_BASED = "project_based"  # Proje bazlı
    COMMISSION = "commission"  # Komisyon
    INVESTMENT = "investment"  # Yatırım getirisi
    OTHER = "other"  # Diğer

class IncomeStatus(str, Enum):
    PLANNED = "planned"      # Planlandı
    INVOICED = "invoiced"    # Faturalandı
    PAID = "paid"           # Ödendi
    OVERDUE = "overdue"     # Gecikmiş
    CANCELLED = "cancelled"  # İptal edildi

class RecurrenceType(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    YEARLY = "yearly"

class IncomeSourceBase(BaseModel):
    name: str = Field(..., description="Gelir kaynağı adı")
    description: Optional[str] = Field(None, description="Açıklama")
    person_id: Optional[str] = Field(None, description="İlgili kişi/kurum ID")
    income_type: IncomeType = Field(..., description="Gelir türü")
    category: str = Field(..., description="Gelir kategorisi")
    expected_amount: float = Field(..., gt=0, description="Beklenen tutar")
    currency: str = Field(default="TRY", description="Para birimi")
    is_recurring: bool = Field(False, description="Tekrarlayan gelir mi")
    recurrence_type: Optional[RecurrenceType] = Field(None, description="Tekrar türü")
    recurrence_interval: Optional[int] = Field(None, description="Tekrar aralığı")
    start_date: datetime = Field(..., description="Başlangıç tarihi")
    end_date: Optional[datetime] = Field(None, description="Bitiş tarihi")
    is_active: bool = Field(True, description="Aktif mi")
    notes: Optional[str] = Field(None, description="Notlar")

class IncomeSourceCreate(IncomeSourceBase):
    pass

class IncomeSourceUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    person_id: Optional[str] = None
    income_type: Optional[IncomeType] = None
    category: Optional[str] = None
    expected_amount: Optional[float] = Field(None, gt=0)
    currency: Optional[str] = None
    is_recurring: Optional[bool] = None
    recurrence_type: Optional[RecurrenceType] = None
    recurrence_interval: Optional[int] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = None

class IncomeSource(IncomeSourceBase):
    id: Optional[str] = Field(default=None, alias="_id")
    total_received: float = Field(0.0, description="Toplam alınan tutar")
    total_expected: float = Field(0.0, description="Toplam beklenen tutar")
    income_count: int = Field(0, description="Gelir sayısı")
    last_income_date: Optional[datetime] = Field(None, description="Son gelir tarihi")
    next_expected_date: Optional[datetime] = Field(None, description="Sonraki beklenen tarih")
    created_by: str = Field(..., description="Oluşturan kullanıcı")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True
        json_encoders = {datetime: lambda v: v.isoformat()}

class IncomeRecordBase(BaseModel):
    income_source_id: str = Field(..., description="Gelir kaynağı ID")
    amount: float = Field(..., gt=0, description="Gelir tutarı")
    currency: str = Field(default="TRY", description="Para birimi")
    description: str = Field(..., description="Gelir açıklaması")
    expected_date: datetime = Field(..., description="Beklenen tarih")
    actual_date: Optional[datetime] = Field(None, description="Gerçekleşen tarih")
    status: IncomeStatus = Field(IncomeStatus.PLANNED, description="Durum")
    invoice_number: Optional[str] = Field(None, description="Fatura numarası")
    transaction_id: Optional[str] = Field(None, description="İlgili işlem ID")
    bank_account_id: Optional[str] = Field(None, description="Banka hesabı ID")
    payment_method: Optional[str] = Field(None, description="Ödeme yöntemi")
    late_fee: float = Field(0.0, description="Gecikme ücreti")
    discount: float = Field(0.0, description="İndirim")
    tax_amount: float = Field(0.0, description="Vergi tutarı")
    net_amount: float = Field(0.0, description="Net tutar")
    notes: Optional[str] = Field(None, description="Notlar")

class IncomeRecordCreate(IncomeRecordBase):
    pass

class IncomeRecordUpdate(BaseModel):
    amount: Optional[float] = Field(None, gt=0)
    currency: Optional[str] = None
    description: Optional[str] = None
    expected_date: Optional[datetime] = None
    actual_date: Optional[datetime] = None
    status: Optional[IncomeStatus] = None
    invoice_number: Optional[str] = None
    transaction_id: Optional[str] = None
    bank_account_id: Optional[str] = None
    payment_method: Optional[str] = None
    late_fee: Optional[float] = None
    discount: Optional[float] = None
    tax_amount: Optional[float] = None
    net_amount: Optional[float] = None
    notes: Optional[str] = None

class IncomeRecord(IncomeRecordBase):
    id: Optional[str] = Field(default=None, alias="_id")
    created_by: str = Field(..., description="Oluşturan kullanıcı")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True
        json_encoders = {datetime: lambda v: v.isoformat()}

class IncomeSourceSummary(BaseModel):
    """Gelir kaynağı özeti"""
    id: str
    name: str
    income_type: IncomeType
    category: str
    expected_amount: float
    currency: str
    total_received: float
    total_expected: float
    income_count: int
    last_income_date: Optional[datetime]
    next_expected_date: Optional[datetime]
    status: str  # "active", "overdue", "upcoming"
    person_name: Optional[str]
    completion_rate: float  # Gerçekleşme oranı

class IncomeRecordSummary(BaseModel):
    """Gelir kaydı özeti"""
    id: str
    income_source_name: str
    amount: float
    currency: str
    description: str
    expected_date: datetime
    actual_date: Optional[datetime]
    status: IncomeStatus
    invoice_number: Optional[str]
    days_overdue: Optional[int]
    person_name: Optional[str]

class IncomeStatistics(BaseModel):
    """Gelir istatistikleri"""
    total_sources: int = Field(0, description="Toplam gelir kaynağı")
    active_sources: int = Field(0, description="Aktif gelir kaynakları")
    recurring_sources: int = Field(0, description="Tekrarlayan gelir kaynakları")
    total_expected_this_month: float = Field(0.0, description="Bu ay beklenen toplam")
    total_received_this_month: float = Field(0.0, description="Bu ay alınan toplam")
    total_overdue: float = Field(0.0, description="Toplam gecikmiş")
    overdue_count: int = Field(0, description="Gecikmiş gelir sayısı")
    upcoming_week: float = Field(0.0, description="Gelecek hafta beklenen")
    upcoming_month: float = Field(0.0, description="Gelecek ay beklenen")
    top_sources: List[Dict[str, Any]] = Field(default_factory=list, description="En büyük gelir kaynakları")
    monthly_trends: List[Dict[str, Any]] = Field(default_factory=list, description="Aylık trendler")
    by_category: List[Dict[str, Any]] = Field(default_factory=list, description="Kategoriye göre dağılım")
    by_status: List[Dict[str, Any]] = Field(default_factory=list, description="Duruma göre dağılım")

class IncomeProjection(BaseModel):
    """Gelir projeksiyonu"""
    month: str = Field(..., description="Ay (YYYY-MM)")
    expected_total: float = Field(0.0, description="Beklenen toplam")
    confirmed_total: float = Field(0.0, description="Kesinleşen toplam")
    recurring_total: float = Field(0.0, description="Tekrarlayan toplam")
    one_time_total: float = Field(0.0, description="Tek seferlik toplam")
    confidence_level: float = Field(0.0, description="Güven seviyesi (0-100)")

class IncomeForecast(BaseModel):
    """Gelir tahmini"""
    next_6_months: List[IncomeProjection] = Field(default_factory=list, description="Sonraki 6 ay")
    yearly_projection: float = Field(0.0, description="Yıllık projeksiyon")
    growth_rate: float = Field(0.0, description="Büyüme oranı")
    seasonality_factor: float = Field(1.0, description="Mevsimsellik faktörü")
    risk_level: str = Field("medium", description="Risk seviyesi")  # low, medium, high